### DISCUSSION TITLE
Rotating a rotatble bond

### DISCUSSION THREAD

**Discussion by YoavShamir5** [rdkit/rdkit#6710] [Q&A]
Hi, 

Regarding this request: https://sourceforge.net/p/rdkit/mailman/rdkit-discuss/thread/5234E914.4050109%40gmail.com/#msg31404453.
I wonder - has this been implemented in Rdkit? Is there a way to change a dihedral angle of a 3D conformer, keeping one specific part of the molecule fixed (coordinates fixed), and rotating the other part with respect to the first one? Thanks!

**Comment by YoavShamir5** [comment]
Adding a similar question - does Rdkit have a built-in function to rotate a conformer along a bond axis (or any user-specified axis, not the cartesian x/y/z)?

**Comment by ptosco** [comment]
> Regarding this request: https://sourceforge.net/p/rdkit/mailman/rdkit-discuss/thread/5234E914.4050109%40gmail.com/#msg31404453. I wonder - has this been implemented in Rdkit? Is there a way to change a dihedral angle of a 3D conformer, keeping one specific part of the molecule fixed (coordinates fixed), and rotating the other part with respect to the first one? Thanks!

There are API functions to get/set dihedrals:
[SetDihedralDeg](http://rdkit.org/docs/source/rdkit.Chem.rdMolTransforms.html?#rdkit.Chem.rdMolTransforms.SetDihedralDeg)
[GetDihedralDeg](http://rdkit.org/docs/source/rdkit.Chem.rdMolTransforms.html?#rdkit.Chem.rdMolTransforms.GetDihedralDeg)
[SetDihedralRad](https://rdkit.org/docs/source/rdkit.Chem.rdMolTransforms.html?#rdkit.Chem.rdMolTransforms.SetDihedralRad)
[GetDihedralRad](https://rdkit.org/docs/source/rdkit.Chem.rdMolTransforms.html?#rdkit.Chem.rdMolTransforms.GetDihedralRad)

> Adding a similar question - does Rdkit have a built-in function to rotate a conformer along a bond axis (or any user-specified axis, not the cartesian x/y/z)?

There isn't, but it is straightforward to do by combining 3D transformations. You only need to (1) translate the conformer such that the origin coincides with the bond's `BeginAtom` (2) compute the angle between the `BeginAtom-EndAtom` vector and the X axis and rotate the conformer by that angle. then (3) rotate the conformer around the X axis by the desired amount, Finally, apply the inverse rotation (2) and translation (1) to restore the original location of the molecule in Cartesian space.

**Reply by YoavShamir5** [reply]
Thanks for the info! I am aware of the set/get dihedral functions, but I am not sure if it is possible to define that one part (connected to one side of the rotating bond) will stay rigid and only the other one will rotate.

**Reply by ptosco** [reply]
That's explained in the docstring:
"Sets the dihedral angle in degrees between atoms `i`, `j`, `k`, `l`; all atoms bonded to atom `l` are moved"

**Reply by YoavShamir5** [reply]
Missed that part! thanks a lot
